vm/src/exceptions.rs

```diff
@@ -522,13 +522,84 @@ pub fn create_exception_type(name: &str, base: &PyTypeRef) -> PyTypeRef {
     create_type_with_slots(name, PyType::static_type(), base, exception_slots())
 }
 
+macro_rules! extends_exception {
+    // This case is like `SimpleExtendsException`:
+    (
+        $class_name: ident,
+        $class_name_literal: literal,
+        $base_class: ident,
+        $base_class_literal: literal,
+        $docs: tt
+    ) => {
+        simple_extends_exception!(
+            $class_name,
+            $class_name_literal,
+            $base_class,
+            $base_class_literal,
+            $docs,
+            {}
+        );
+    };
+
+    // TODO: Do we need `MiddlingExtendsException`?
+
+    // This case is like `ComplexExtendsException`:
+    (
+        $class_name: ident,
+        $class_name_literal: literal,
+        $base_class: ident,
+        $base_class_literal: literal,
+        $docs: tt,
+        { $($attr_name:expr => $attr_value:expr),* $(,)* }
+    ) => {
+        #[pyclass(module = false, name = $class_name_literal, base = $base_class_literal)]
+        #[derive(Debug)]
+        struct $class_name {}
+
+        #[pyimpl(flags(BASETYPE))]
+        impl $class_name {
+            // TODO: we probably need to add `new` as well
+
+            #[pymethod(magic)]
+            fn init(
+                zelf: PyRef<PyBaseException>,
+                args: FuncArgs,
+                vm: &VirtualMachine,
+            ) -> PyResult<()> {
+                // We need this method, because of how `CPython` copies `init`
+                // from `BaseException` in `SimpleExtendsException` macro.
+                // See: `(initproc)BaseException_init`
+                // In: https://github.com/python/cpython/blob/main/Objects/exceptions.c
+                $base_class::init(zelf, args, vm)
+            }
+
+            #[extend_class]
+            fn extend_class_with_fields(ctx: &PyContext, class: &PyTypeRef) {
+                class.set_str_attr("__doc__", ctx.new_str($docs.to_owned()));
+                $(
+                    class.set_str_attr($attr_name, $attr_value);
+                )*
+            }
+        }
+    };
+}
+
+extends_exception! {
+    PyKeyboardInterrupt,
```

Comment:
What I don't like here is that I have to repeat `ident` and `literal` names. Why?
- I cannot use just `ident` names with `#[pyclass(name = ..., base = ...)]`, because they explicitly need `literal` strings. I even cannot use `stringify!`, because it is not expanded in `#[pyclass]` attributes
- I cannot use just `literal`, because Rust cannot `unstrigify!` things, unless https://docs.rs/unstringify/0.1.1/unstringify/macro.unstringify.html is installed

Reply:
Friendly ping @youknowone @DimitrisJim

Reply:
I don't any a nice idea for this. This problem is known to be a tranditional rust macro problem. So it looks just ok to me for now. Will it be better if pyclass takes both ident and str for its macro arguments? @coolreader18

Reply:
I think so, `#[pyclass(name = :ident | :literal, base = :ident | :literal)]` seems like a good solution 👍 
I will implement it in a separate PR.

Reply:
Looks like that there's nothing I can do. It is a limitation of `pub type AttributeArgs = Vec<NestedMeta>;`, where `NestedMeta` can be:

```rust
    /// ## Path
    ///
    /// A meta path is like the `test` in `#[test]`.
    ///
    /// ## List
    ///
    /// A meta list is like the `derive(Copy)` in `#[derive(Copy)]`.
    ///
    /// ## NameValue
    ///
    /// A name-value meta is like the `path = "..."` in `#[path =
    /// "sys/windows.rs"]`.
```

And `NameValue` only allows `Lit` values 😞 
So, I guess that the only way is to pass `ident` and `literal` as two separate params.

Reply:
https://github.com/dtolnay/syn/blob/5a9efb0b772b583af93cdca9453fa9fd4e522641/src/lit.rs#L745

Reply:
Hm `#[pyexception($class_name, $base_class)]` seem to work, I will continue my experiments 👍